Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp

```diff
@@ -314,6 +320,75 @@ MediaStreamTrack::TrackCapabilities MediaStreamTrack::getCapabilities() const
     return result;
 }
 
+void MediaStreamTrack::queueAndProcessSerialAction(Function<Ref<GenericPromise>(SerialActionResolution)>&& action)
+{
+    ASSERT(isMainThread());
+    m_pendingActionQueue.append(WTFMove(action));
+    processNextSerialAction();
+}
+
+void MediaStreamTrack::processNextSerialAction()
+{
+    ASSERT(isMainThread());
+    if (m_currentSerialAction || m_pendingActionQueue.isEmpty())
+        return;
+
+    m_currentSerialAction = { m_pendingActionQueue.takeFirst() };
+    m_currentSerialAction.value()(SerialActionResolution::Run)->whenSettled(RunLoop::main(), [protectedThis = Ref { *this }] {
+        protectedThis->m_currentSerialAction = { };
+        protectedThis->processNextSerialAction();
+    });
+}
+
+auto MediaStreamTrack::takePhoto(PhotoSettings&& settings) -> Ref<TakePhotoPromise>
+{
+    TakePhotoPromise::Producer producer;
+    Ref<TakePhotoPromise> promise = producer;
+
+    queueAndProcessSerialAction([settings = WTFMove(settings), protectedThis = Ref { *this }, producer = WTFMove(producer)](SerialActionResolution resolution) mutable -> Ref<GenericPromise> {
```

Comment:
It might be good to explain in the commit message why we are introducing a queue.